keras/layers/core.py

```diff
@@ -276,20 +276,33 @@ def get_config(self):
 
 
 class Merge(Layer):
-    def __init__(self, layers, mode='sum', concat_axis=-1):
+    def __init__(self, layers, mode='sum', concat_axis=-1,dot_axes=-1):
         ''' Merge the output of a list of layers or containers into a single tensor.
             mode: {'sum', 'mul', 'concat', 'ave', 'join'}
         '''
         if len(layers) < 2:
             raise Exception("Please specify two or more input layers (or containers) to merge")
 
-        if mode not in {'sum', 'mul', 'concat', 'ave', 'join'}:
+        if mode not in {'sum', 'mul', 'concat', 'ave', 'join', 'cos', 'dot'}:
             raise Exception("Invalid merge mode: " + str(mode))
 
-        if mode in {'sum', 'mul', 'ave'}:
+        if mode in {'sum', 'mul', 'ave', 'cos'}:
             input_shapes = set([l.output_shape for l in layers])
             if len(input_shapes) > 1:
                 raise Exception("Only layers of same output shape can be merged using " + mode + " mode")
+        if mode in {'cos', 'dot'}:
+            if len(layers) > 2:
+                raise Exception(mode + " merge takes exactly 2 layers")
+            shape1 = layers[0].output_shape
+            shape2 = layers[1].output_shape
+            if mode == 'dot':
+                if type(dot_axes) == int:
+                    if dot_axes < 0:
+                        dot_axes = len(shape1) - 1
+                    dot_axes = [range(len(shape1) - dot_axes, len(shape2)), range(1,dot_axes + 1)]
+                for i in rage(len(dot_axes[0])):
```

Comment:
Typo: it's `range`. This means that this mode wasn't tested, and in fact, never run.